#!/bin/bash

##################################
##################################
####     ESTIMATING ALLELE    ####
####  FREQUENCIES WITH ANGSD  ####
##################################
##################################

#####################
#### PBS DETAILS ####
#####################

#### User details

#### Project name
#PBS -P rh35
####  Job name
#PBS -N SnakemakeAlleleFrequencies
#### Email for notification
#PBS -M oliver.stuart@anu.edu.au
#### Only email if aborted or complete
#PBS -m ae

#### Resource requirements

#### How many processors
#PBS -l ncpus=48
#### How much memory all up
#PBS -l mem=192GB
#### How much walltime
#PBS -l walltime=48:00:00
#### How much local space required
#PBS -l jobfs=10GB
#### Start the job by moving to the directory from which it was submitted
#PBS -l wd

#### STDOUT/ERR redirection

#### STDOUT
#PBS -o /home/147/os4776/SnakemakeAlleleFrequencies.out
#### STDERR
#PBS -e /home/147/os4776/SnakemakeAlleleFrequencies.err

###################
#### ENV SETUP ####
###################

# Data and script directories
SCRATCH=/scratch/rh35/os4776/LHISI_WGS
REF_DIR=${SCRATCH}/References
ALN_DIR=${SCRATCH}/Alignments
SCR_DIR=${SCRATCH}/Scripts
HOME_DIR=/home/147/os4776

#mkdir=${SCRATCH}/Analyses/AlleleFrequencies

OUT_DIR=${SCRATCH}/Analyses/AlleleFrequencies

# Local software directories
ANGSD=/home/147/os4776/angsd/angsd
SNAKEMAKE=${HOME_DIR}/.local/bin/snakemake

# Get into working directory
WORKING_DIR=$PBS_JOBFS
cd ${WORKING_DIR}

# Make output directory for copying later
#DATE=$(date '+%d%m%Y_%H%M%S')
#mkdir ${SCRATCH}/Analyses/${PBS_JOBNAME}_${DATE}
#OUT_DIR=${SCRATCH}/Analyses/${PBS_JOBNAME}_${DATE}
# For this, we don't need to do this
# Maybe retiring this storage strategy in future

# Load python module to allow snakemake to run
module load python3/3.10.4

#######################
#### ACTUAL SCRIPT ####
#######################

# Make a file containing the names of all individuals to be included
# Remove samples that didn't run well

find $ALN_DIR | grep bam$ | grep -v "C01216\|C01233\|C01218\|C01226\|C10133" > all_bam_list.txt

# Now make individual files for the populations

grep "PAUL\|VAN" all_bam_list.txt > wild_bam_list.txt
grep "C01230\|C01225\|C10223\|C01211\|C01217\|C01220\|C01234\|C01224" all_bam_list.txt > lhip_bam_list.txt
grep "C01210\|C01215\|C01223\|C01222\|C01232\|C01231\|C01219\|C01227\|C01213" all_bam_list.txt > lhisi_bam_list.txt

# Now run the snakefile

${SNAKEMAKE} -s ${HOME_DIR}/SnakefileAlleleFrequencies --cores 42

#${HOME_DIR}/.local/bin/snakemake \
#-s ${HOME_DIR}/SnakefileGadiTest \
#--cores 8 \
#--cluster "qsub -A rh35 -l mem={resources.mem},walltime={resources.time},jobfs={resources.tempstorage},ncpus={threads} -N {wildcards.sample}_{rule} -k eo -o /home/147/os4776/{wildcards.sample}_{rule}.out -e /home/147/os4776/{wildcards.sample}_{rule}.err" \
#--rerun-incomplete --notemp --nolock --max-jobs-per-second 1 --jobs 10
